### PR TITLE
Initial structure for more data-driven actions

### PR DESCRIPTION
**Pull request by nikolaj-a**
Define an initial interface for actions, as well an implementation and data structures for simple actions without any check involved and actions involving a single check.

### Simple Action Examples

**Take Cover**
```js
const action = new SimpleAction({
  cost: 1,
  description: "You press yourself against a wall or duck behind an obstacle to take better advantage of cover. If you would have standard cover, you instead gain greater cover, which provides a +4 circumstance bonus to AC; to Reflex saves against area effects; and to Stealth checks to Hide, Sneak, or otherwise avoid detection. Otherwise, you gain the benefits of standard cover (a +2 circumstance bonus instead). This lasts until you move from your current space, use an attack action, become unconscious, or end this effect as a free action.",
  effect: "Compendium.pf2e.other-effects.I9lfZUiCwMiGogVi", // Effect: Cover
  name: "Take Cover",
  requirements: "You are benefiting from cover, are near a feature that allows you to take cover, or are prone.",
});

action.use({ event }).catch(reason => ui.notifications.error(reason));
```
Take Cover Chat Card

**Raise a Shield**
```js
const action = new SimpleAction({
  cost: 1,
  description: "You position your shield to protect yourself. When you have Raised a Shield, you gain its listed circumstance bonus to AC. Your shield remains raised until the start of your next turn.",
  effect: "Compendium.pf2e.equipment-effects.2YgXoHvJfrDHucMr", // Effect: Raise a Shield
  name: "Raise a Shield",
  requirements: "You are wielding a shield.",
});

action.use({ event }).catch(reason => ui.notifications.error(reason));
```
Raise a Shield Chat Card

### Single Check Action Examples

**Treat Poison**
```js
const action = new SingleCheckAction({
  cost: 1,
  description: "You treat a patient to prevent the spread of poison. Attempt a Medicine check against the poison's DC. After you attempt to Treat a Poison for a creature, you can't try again until after the next time that creature attempts a save against the poison.",
  name: "Treat Poison",
  notes: [
    { outcome: ["criticalSuccess"], text: "You grant the creature a +4 circumstance bonus to its next saving throw against the poison.", title: "Critical Success" },
    { outcome: ["success"], text: "You grant the creature a +2 circumstance bonus to its next saving throw against the poison.", title: "Success" },
    { outcome: ["criticalFailure"], text: "Your efforts cause the creature to take a -2 circumstance penalty to its next save against the poison.", title: "Critical Failure" },
  ],
  requirements: "You are holding healer's tools, or you are wearing them and have a hand free",
  statistic: "medicine",
  traits: ["manipulate"],
});

action.use({
  difficultyClass: { label: "Giant Centipede Venom DC", value: 17 },
  event,
}).catch(reason => ui.notifications.error(reason));
```
Treat Poison Chat Card

**Trip**
```js
const action = new SingleCheckAction({
  cost: 1,
  description: "You try to knock a creature to the ground. Attempt an Athletics check against the target's Reflex DC.",
  difficultyClass: "saves.reflex",
  name: "Trip",
  notes: [
    { outcome: ["criticalSuccess"], text: "The target falls and lands prone and takes 1d6 bludgeoning damage.", title: "Critical Success" },
    { outcome: ["success"], text: "The target falls and lands prone.", title: "Success" },
    { outcome: ["criticalFailure"], text: "You lose your balance and fall and land prone.", title: "Critical Failure" },
  ],
  requirements: "You have at least one hand free. Your target can't be more than one size larger than you.",
  statistic: "athletics",
  traits: ["attack"],
});

action.use({ event }).catch(reason => ui.notifications.error(reason));
```
Trip Chat Card

**Administer First Aid - Stop Bleeding**
```js
const action = new game.pf2e.SingleCheckAction({
  cost: 2,
  description: "<p>You perform first aid on an adjacent creature that is dying or bleeding. If a creature is both dying and bleeding, choose which ailment you're trying to treat before you roll. You can Administer First Aid again to attempt to remedy the other effect.</p><ol><li><strong>Stabilize</strong> Attempt a Medicine check on a creature that has 0 Hit Points and the dying condition. The DC is equal to 5 + that creature's recovery roll DC (typically 15 + its dying value).</li><li><strong>Stop Bleeding</strong> Attempt a Medicine check on a creature that is taking persistent bleed damage, giving them a chance to make another flat check to remove the persistent damage. The DC is usually the DC of the effect that caused the bleed.</li></ol><p><strong>Success</strong> If you're trying to stabilize, the creature loses the dying condition (but remains unconscious). If you're trying to stop bleeding, the creature attempts a flat check to end the bleeding.<br><strong>Critical Failure</strong> If you were trying to stabilize, the creature's dying value increases by 1. If you were trying to stop bleeding, it immediately takes an amount of damage equal to its persistent bleed damage.<p>",
  name: "Administer First Aid",
  requirements: "You are holding healer's tools, or you are wearing them and have a hand free",
  statistic: "medicine",
  traits: ["manipulate"],
  variants: [
    {
      name: "Stabilize",
      notes: [{
        outcome: ["criticalSuccess", "success"],
        text: "The creature loses the dying condition (but remains unconscious).",
        title: "Success",
      }, {
        outcome: ["criticalFailure"],
        text: "The creature's dying value increases by 1.",
        title: "Critical Failure",
      }],
    }, {
      name: "Stop Bleeding",
      notes: [{
        outcome: ["criticalSuccess", "success"],
        text: "The creature attempts a flat check to end the bleeding.",
        title: "Success",
      }, {
        outcome: ["criticalFailure"],
        text: "The creature immediately takes an amount of damage equal to its persistent bleed damage.",
        title: "Critical Failure",
      }],
    },
  ],
});

action.use({
  difficultyClass: { label: "Serrating Rune Bleed DC", value: 28 },
  event,
  variant: "stop-bleeding",
}).catch(reason => ui.notifications.error(reason));
```
Administer First Aid - Stop Bleeding Chat Card

### CODE REVIEW

File: src/module/system/actions/base.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+interface BaseActionData {
+    description?: string;
+    glyph?: ActionGlyph;
+    name: string;
+    traits?: string[];
+}
+
+abstract class BaseAction implements Action {
+    readonly #description?: string;
+    readonly #glyph?: ActionGlyph;
```
Comment (stwlam) [inline src/module/system/actions/base.ts:13]: Storing a glyph is a little inline-brained. It should take an action cost, and a glyph derived from that (or not for exploration/downtime activities).

File: src/module/system/actions/base.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    variants: { 0: ActionVariantDataType } & ActionVariantDataType[];
+}
+
+abstract class BaseActionVariant implements ActionVariant {
+    readonly #action: BaseAction<BaseActionVariantData, BaseActionVariant>;
+    readonly #cost?: ActionCost;
+    readonly #description?: string;
+    readonly #name?: string;
+    readonly #slug?: string;
+    readonly #traits?: string[];
```
Comment (stwlam) [inline src/module/system/actions/base.ts:32]: I don't know that we need this level of paranoia. The `readonly` type is sufficient to keep a contributor from trying to reassign them, and if someone is being silly in a macro, then that's on them.

Comment (stwlam) [reply src/module/system/actions/base.ts:32]: But the getters don't serve any purpose.

**Comment by stwlam** [comment]
Can we move all this to `@actor/action`?

**Comment by nikolaj-a** [comment]
> Can we move all this to `@actor/action`?

Sure, I don't see why not.

**Pull request merged**